### PR TITLE
Add rounded corners on the bar chart #1066

### PR DESCRIPTION
**Pull request by Neral**
Added ability to draw rounded corners on each bar.

By default rounded corners is disabled and to enable it `isDrawRoundedBarEnabled` should be used.
Also using `barRoundingCorners` could be decided which corners should be rounded ([reference](https://developer.apple.com/reference/uikit/uirectcorner)).

Feature: #1066 

charts

### CODE REVIEW

**Comment by liuxuan30** [comment]
A corner case would be, if the bar is very thin, the rounded bar looks like a thin oval.. That's what I am not satisfied with the CG API :)

**Comment by arronhunt** [comment]
This is great, would be awesome if we could get these for candlestick charts as well :)

**Comment by vinodha88** [comment]
Any idea when the rounded corner feature will be merged ? I am using cocoapods (pod 'Charts') in my app and wanted to check this patch .

**Comment by liuxuan30** [comment]
No ETA, as @danielgindi is quite busy recently.
Besides, such feature needs to be considered for Android, so it usually take a very long time.

For those want to try out, you should use source code instead of carthage, pod. More flexible.

**Comment by acegreen** [comment]
@Neral building on top of your work, I made cornerRadius a user input https://github.com/omsignal/Charts

**Comment by acegreen** [comment]
Noticed that we need to consider when the bar is stacked, currently it rounds all entries

**Comment by acegreen** [comment]
@Neral Also I noticed if the axisMinimum is not set to 0, the bottom corners don't show

**Comment by gsolanki1509** [comment]
i couldn't find "isDrawRoundedBarEnabled" property for Bar chart.

I'm using Swift 3.0. Please let me know

**Comment by liuxuan30** [comment]
this has not been merged yet. and why @gsolanki1509 you are reviewerers...

File: Source/Charts/Renderers/BarChartRenderer.swift [hunk 1]
```diff
@@ -310,14 +310,36 @@ open class BarChartRenderer: BarLineScatterCandleBubbleRenderer
...
-            
-            if drawBorder
+
+            if dataProvider.isDrawRoundedBarEnabled
             {
-                context.setStrokeColor(borderColor.cgColor)
-                context.setLineWidth(borderWidth)
-                context.stroke(barRect)
+                let cornerRadius = CGSize(width: barRect.width / 2.0, height: barRect.width / 2.0)
+                #if os(OSX)
```
Comment (liuxuan30) [inline Source/Charts/Renderers/BarChartRenderer.swift]: Just FYI, I came across API like `CGPathCreateWithRoundedRect()` and `CGPathAddRoundedRect()`, looks like we can replace the UIBezierPath APIs?

File: Source/Charts/Renderers/BarChartRenderer.swift [hunk 2]
```diff
@@ -310,14 +310,36 @@ open class BarChartRenderer: BarLineScatterCandleBubbleRenderer
...
+                    let bezierPath = NSBezierPath(roundedRect: barRect, xRadius: cornerRadius.width, yRadius: cornerRadius.height)
```
Comment (liuxuan30) [inline Source/Charts/Renderers/BarChartRenderer.swift]: Just FYI, I came across API like `CGPathCreateWithRoundedRect()` and `CGPathAddRoundedRect()`, looks like we can replace the UIBezierPath APIs?

File: Source/Charts/Renderers/BarChartRenderer.swift [hunk 3]
```diff
@@ -310,14 +310,36 @@ open class BarChartRenderer: BarLineScatterCandleBubbleRenderer
...
-            context.fill(barRect)
-            
-            if drawBorder
+
+            if dataProvider.isDrawRoundedBarEnabled
             {
-                context.setStrokeColor(borderColor.cgColor)
-                context.setLineWidth(borderWidth)
-                context.stroke(barRect)
+                let cornerRadius = CGSize(width: barRect.width / 2.0, height: barRect.width / 2.0)
```
Comment (nickskull) [inline Source/Charts/Renderers/BarChartRenderer.swift]: It would be nice if this was a property instead, so that we can specify a custom corner radius.

**Comment by Kit-OS** [comment]
It would be great that someone could merge this feature soon :)

**Comment by vivianegon00** [comment]
is the rounded bars option coming soon? Like on the Apple's activity tracker app?

**Comment by liuxuan30** [comment]
@vivianegon00 I think it's different. It's just bars with rounded corner.

**Comment by vivianegon00** [comment]
@liuxuan30 that's exactly what I need, bars with rounded corners... this option is not available now

**Comment by kacperios** [comment]
anyone has information when this feature will be available ?

**Comment by weet1** [comment]
@liuxuan30 a lot of people need need this function, is there any hope that it will merge? :)

**Comment by liuxuan30** [comment]
I have said long time ago in another thread, though it looks nice, but when bar becomes thinner, it will looks like oval instead of rounded bars, so kind of flaw. I'm not sure if there are any better solutions, so this is postponed. you guys are ok with thin bars?

**Comment by juangz** [comment]
This would be awesome to have!!! Needing this asap!!

**Comment by jjatie** [comment]
@liuxuan30 Looks like this is a heavily desired feature, and the author has abandoned the PR. Shall I reimplement this? I think the thin bar issue is fine. If a consumer doesn't like it they can turn off rounded corners.

**Comment by jjatie** [comment]
@liuxuan30 Looks like this is a heavily desired feature, and the author has abandoned the PR. Shall I reimplement this? I think the thin bar issue is fine. If a consumer doesn't like it they can turn off rounded corners.

**Comment by liuxuan30** [comment]
@jjatie if thin bar is acceptable, then we can move forward. But I am still struggling with it, as when the bars become more, it will more like an oval, which is not ideal.

We could first write a solid and flexible code base for this, and if we have better solutions, we can easily switch them. Don't be limited at my raw implementation about `UIBezierPath(roundedRect: barRect, byRoundingCorners: dataSet.barRoundingCorners, cornerRadii: cornerRadius)`, maybe we could take a look at `CGPathCreateWithRoundedRect()` and `CGPathAddRoundedRect()`
and other possible solution.

File: Source/Charts/Charts/BarChartView.swift [hunk 4]
```diff
@@ -180,4 +194,7 @@ open class BarChartView: BarLineChartViewBase, BarChartDataProvider
     
     /// - returns: `true` if drawing shadows (maxvalue) for each bar is enabled, `false` ifnot
     open var isDrawBarShadowEnabled: Bool { return drawBarShadowEnabled }
+
+    /// - returns: `true` if drawing rounded bars is enabled, `false` ifnot
```
Comment (diegopetrucci) [inline Source/Charts/Charts/BarChartView.swift]: `ifnot` typo

File: Source/Charts/Charts/CombinedChartView.swift [hunk 5]
```diff
@@ -222,4 +229,7 @@ open class CombinedChartView: BarLineChartViewBase, CombinedChartDataProvider
     
     /// - returns: `true` the highlight is be full-bar oriented, `false` ifsingle-value
     open var isHighlightFullBarEnabled: Bool { return highlightFullBarEnabled }
+
+    /// - returns: `true` if drawing rounded bars is enabled, `false` ifnot
```
Comment (diegopetrucci) [inline Source/Charts/Charts/CombinedChartView.swift]: `ifnot` typo

**Comment by mikestalker** [comment]
One year in review? Really?

**Comment by liuxuan30** [comment]
@mikestalker it's open source and you have all the code you need. Either you could merge yourself or wait. At least you didn't pay me/us, and we all have a job to do. 

Besides I have said the implementation I found is not satisfying. You can contribute a better one rather than complain.

**Comment by ravipxm** [comment]
@liuxuan30 any ETA on this?

**Comment by xRena** [comment]
+1

**Comment by acegreen** [comment]
@danielgindi @liuxuan30  almost 2 years in review?

**Comment by mikestalker** [comment]
@acegreen , @liuxuan30 already answered (=

> @mikestalker  it's open source and you have all the code you need. Either you could  merge yourself or wait. At least you didn't pay me/us to do this, and we  all have a job to do. Besides I have said the implementation I found is not satisfying. You can contribute a better one rather than complain.

**Comment by liuxuan30** [comment]
Until a better solution emerge. As I said long time ago, this solution is not perfect (more bars, the oval become sharp). I don't think this is worth merging yet. I keep this open is some people could live with the defect. Use at your own risk

**Comment by jjatie** [comment]
#3754

**Pull request closed without merging**